### PR TITLE
Move iTerm-server daemon to per-user namespace. Issue 4147

### PR DESCRIPTION
**Pull request by pradorocchi**
This PR fixes issue 4147

### CODE REVIEW

File: sources/shell_launcher.c [hunk 1]
```diff
@@ -140,6 +143,102 @@ sudo_closefrom(int lowfd) {
...
+   ----
+       System_Namespace
+            Per-User_Namespace
+                Per-Session_Namespace(GUI WindowServer)
+   ----
+   Technically the GUI per-session namespace is called 'Aqua' Session by Apple's API docs.
+
+   Issue 4147 happens because the forked daemon is running in the Aqua session namespace. So
+   when the user logout or a GUI crash happens, the forked daemon remains isolated in a lost
+   per-session namespace. It partially looses it's IPC capabilities to comunicate with other
```
Comment (ooglek) [inline sources/shell_launcher.c]: > partially _loses_ it's ... to _communicate_ ...

File: sources/shell_launcher.c [hunk 2]
```diff
@@ -140,6 +143,102 @@ sudo_closefrom(int lowfd) {
...
+   where it is running to a level above it: To the 'per-user' namespace. This is necessary to
+   keep the daemon acessible and securelly working when the user logout from the GUI session,
+   or if the user changes the session via FastUserSwitching, or in case of crashes of the GUI
+   session. (In a crash, WindowServer will recreate a new per-session namespace).
+
+   By moving iTerm-Server Daemon we make it be available as a service to any 'per-session'
+   namespace that already exists [the current one], or any new 'per-session' namespaces that
+   may in future be (re)created below it.
+
+   Important is to mention that it's availability is to serve just the same user (the same UID).
```
Comment (ooglek) [inline sources/shell_launcher.c]: > that _its_ availability ...

File: sources/shell_launcher.c [hunk 3]
```diff
@@ -140,6 +143,102 @@ sudo_closefrom(int lowfd) {
...
+   This is what we want, because all other BSD preparations here is done to create the
+   iTerm-Server to serve just that given user who started it. If another user is logged-in via
+   GUI via fast user switching, this other user will have it's own Per-User session, and will
```
Comment (ooglek) [inline sources/shell_launcher.c]: > have _its_ own ...

File: sources/shell_launcher.c [hunk 4]
```diff
@@ -140,6 +143,102 @@ sudo_closefrom(int lowfd) {
...
+   have below it his Aqua per-session GUI belonging to it's UID, and his programs running there,
```
Comment (ooglek) [inline sources/shell_launcher.c]: > to _its_ UID ...

File: sources/shell_launcher.c [hunk 5]
```diff
@@ -140,6 +143,102 @@ sudo_closefrom(int lowfd) {
...
+   including his own iTerm-Server daemons forked just for it's UID use.
```
Comment (ooglek) [inline sources/shell_launcher.c]: > for _its_ UID ...

File: sources/shell_launcher.c [hunk 6]
```diff
@@ -140,6 +143,102 @@ sudo_closefrom(int lowfd) {
...
+   This way iTerm2 daemon is in conformation with current MacOS security rules mechanics,
+   and we completely eliminate all the problems that raised issue 4147. Per-user namespace is
+   the correct place for a user daemon to live and service it's services to the namespaces
+   levels below it.
+
+   *The Per-User namespace is the parent for all other context namespace for a given user. It
+   is never destroyed by GUI logout or crashes. And it will exist until there is at least one
+   daemon or service running on it, for that given user. [* from Apples documentation]
+
+   That's the exact namespace where Launchd places daemons for a given user.
```
Comment (ooglek) [inline sources/shell_launcher.c]: 👍 Great writeup and docs!

File: sources/shell_launcher.c [hunk 7]
```diff
@@ -140,6 +143,102 @@ sudo_closefrom(int lowfd) {
...
+    if (kr == KERN_SUCCESS) {
+        // Detach server daemon process from the user's Aqua 'per-session' namespace,
+        // and move it to the same UID 'per-user' namespace (just one level up).
+        mach_port_mod_refs(mach_task_self(), bootstrap_port, MACH_PORT_RIGHT_SEND, -1);
+        task_set_special_port(mach_task_self(), TASK_BOOTSTRAP_PORT, parent);
+
+        return; /* All Done! now return, as we are ready to begin forking! */
+    }
+    // Never reach. Unless if we had failed to get the parent bootstrap port, which according
+    // to Apple's documentation never happens. But I already saw rare wierd system situations
```
Comment (ooglek) [inline sources/shell_launcher.c]: > saw _weird_ system ...

File: sources/shell_launcher.c [hunk 8]
```diff
@@ -140,6 +143,102 @@ sudo_closefrom(int lowfd) {
...
+    // where it may happen.
+    // Almost impossible, but if it happens, for sure is a BUG. So log it just in case.
+    FDLog(LOG_ERR,"BUG: Error getting parent BS port! Please report this msg and code: '%x'", kr);
```
Comment (ooglek) [inline sources/shell_launcher.c]: I'm a fan of defensive error logging

**Comment by gnachman** [comment]
This is a fantastic PR! Thank you!

File: sources/shell_launcher.c (hunk 6 above)
Comment (pradorocchi) [reply sources/shell_launcher.c]: Thanks!!

File: sources/shell_launcher.c (hunk 7 above)
Comment (pradorocchi) [reply sources/shell_launcher.c]: Thank you very much for the spelling/grammar changes!
My natural language is Portuguese (I'm from Brazil), some (sometimes a lot of) mistakes happens, :)

File: sources/shell_launcher.c (hunk 8 above)
Comment (pradorocchi) [reply sources/shell_launcher.c]: Me too, learned the importance of that on my last 9 years working as developer on the asterisk telephony platform. There are a lot of similar BUG situations on that code, and naturally I created this function the way I'm accustomed to do. 

Thank you for your appreciation!

**Comment by pradorocchi** [comment]
> This is a fantastic PR! Thank you!

Thank you for creating iTerm2 @gnachman, you deserve the best congratulations!

**Comment by gnachman** [comment]
Awesome! How did you figure this out?

**Comment by pradorocchi** [comment]
Here is all the information sources I used to achieve this PR

.

Apple Excellent technical note about the intrinsic mechanism of BSD + Mach kernel portions
https://developer.apple.com/library/archive/technotes/tn2083/_index.html

Apple Libc source code
https://opensource.apple.com/source/Libc/Libc-1272.200.26/

MacOS launchctl source code 
https://github.com/PureDarwin/XPC/blob/master/src/launchctl/launchctl.c

MacOS 10.14 kernel fork() source code
http://newosxbook.com/src.jl?tree=xnu&file=/bsd/kern/kern_fork.c



**Additional researches looking for use cases at :**

MacOS 10.14.0 Kernel source code
http://newosxbook.com/src.jl?tree=xnu-4903.221.2

Various system packages source code for Mojave 10.14.1 that deals with process execution
https://opensource.apple.com/release/macos-10141.html

**Pull request merged**